knora/excel2xml.py

```diff
@@ -0,0 +1,2127 @@
+            elif resource_restype == "Annotation":
+                resource = make_annotation(**kwargs_resource)
+            elif resource_restype == "LinkObj":
+                resource = make_link(**kwargs_resource)
+
+        ########### case property-row ###########
+        else:  # check_notna(row["prop name"]):
+            # based on the property type, the right function has to be chosen
+            if row.get("prop type") not in proptype_2_function:
+                raise BaseError(f"Invalid prop type for property {row['prop name']} in resource {resource_id}")
+            make_prop_function = proptype_2_function[row["prop type"]]
+
+            # every property contains i elements, which are represented in the Excel as groups of
+            # columns named {i_value, i_encoding, i_res ref, i_permissions, i_comment}. Depending
+            # on the property type, some of these items are NA.
+            # Thus, prepare list of PropertyElement objects, with each PropertyElement containing only
+            # the existing items.
+            property_elements: list[PropertyElement] = []
+            for i in range(1, max_prop_count + 1):
+                value = row[f"{i}_value"]
+                if check_notna(value):
+                    kwargs_propelem = {
+                        "value": value,
+                        "permissions": str(row.get(f"{i}_permissions"))
+                    }
+                    if not check_notna(row.get(f"{i}_permissions")):
+                        raise BaseError(f"Missing permissions for value {value} of property {row['prop name']} "
+                                        f"in resource {resource_id}")
+                    if check_notna(row.get(f"{i}_comment")):
+                        kwargs_propelem["comment"] = str(row[f"{i}_comment"])
+                    if check_notna(row.get(f"{i}_encoding")):
+                        kwargs_propelem["encoding"] = str(row[f"{i}_encoding"])
+
+                    property_elements.append(PropertyElement(**kwargs_propelem))
+
+            # create the property and append it to resource
+            kwargs_propfunc: dict[str, Union[str, list[PropertyElement]]] = {
+                "name": row["prop name"],
+                "calling_resource": resource_id
+            }
+            if make_prop_function in single_value_functions or len(property_elements) == 1:
+                kwargs_propfunc["value"] = property_elements
+            else:
+                kwargs_propfunc["values"] = property_elements
+            if check_notna(row["prop list"]):
+                kwargs_propfunc["list_name"] = str(row["prop list"])
+
+            resource.append(make_prop_function(**kwargs_propfunc))
+
+    # append the resource of the very last iteration of the for loop
+    root.append(resource)
+
+    # write file
+    # ----------
+    write_xml(root, f"{default_ontology}-data.xml")
+    print(f"XML file successfully created at {default_ontology}-data.xml")
```

Comment:
this file is definitely too long. It also does many things at once (data class, util functions, process functions, pipeline of processes to be called from the CLI tool), so it can easily be broken up into multiple files.

Reply:
I agree.